# IT Incident Dataset 
<h2 style="color:#2F5597; font-size:36px;">Stage 1 : Data Understanding</h2>

In [2]:
import pandas as pd

In [4]:
incident_df = pd.read_csv("C:/Users/husna/OneDrive/Documents/ca2-90-Husnain-Yaqoob/data/incident_event_log.csv")

incident_df.head()

,number,incident_state,active,reassignment_count,reopen_count,sys_mod_count,made_sla,caller_id,opened_by,opened_at,...,u_priority_confirmation,notify,problem_id,rfc,vendor,caused_by,closed_code,resolved_by,resolved_at,closed_at
0,INC0000045,New,True,0,0,0,True,Caller 2403,Opened by 8,29/2/2016 01:16,...,False,Do Not Notify,?,?,?,?,code 5,Resolved by 149,29/2/2016 11:29,5/3/2016 12:00
1,INC0000045,Resolved,True,0,0,2,True,Caller 2403,Opened by 8,29/2/2016 01:16,...,False,Do Not Notify,?,?,?,?,code 5,Resolved by 149,29/2/2016 11:29,5/3/2016 12:00
2,INC0000045,Resolved,True,0,0,3,True,Caller 2403,Opened by 8,29/2/2016 01:16,...,False,Do Not Notify,?,?,?,?,code 5,Resolved by 149,29/2/2016 11:29,5/3/2016 12:00
3,INC0000045,Closed,False,0,0,4,True,Caller 2403,Opened by 8,29/2/2016 01:16,...,False,Do Not Notify,?,?,?,?,code 5,Resolved by 149,29/2/2016 11:29,5/3/2016 12:00
4,INC0000047,New,True,0,0,0,True,Caller 2403,Opened by 397,29/2/2016 04:40,...,False,Do Not Notify,?,?,?,?,code 5,Resolved by 81,1/3/2016 09:52,6/3/2016 10:00


## Dataset Preview

The first few rows of the dataset were displayed to get an initial understanding of how the data looks.

This helps identify the type of information available in the dataset and gives a starting point for further analysis.

## Dataset Size

In [5]:
incident_df.shape

(141712, 36)

## Column Overview

In [6]:
incident_df.columns.tolist()

['number',
 'incident_state',
 'active',
 'reassignment_count',
 'reopen_count',
 'sys_mod_count',
 'made_sla',
 'caller_id',
 'opened_by',
 'opened_at',
 'sys_created_by',
 'sys_created_at',
 'sys_updated_by',
 'sys_updated_at',
 'contact_type',
 'location',
 'category',
 'subcategory',
 'u_symptom',
 'cmdb_ci',
 'impact',
 'urgency',
 'priority',
 'assignment_group',
 'assigned_to',
 'knowledge',
 'u_priority_confirmation',
 'notify',
 'problem_id',
 'rfc',
 'vendor',
 'caused_by',
 'closed_code',
 'resolved_by',
 'resolved_at',
 'closed_at']

The dataset contains 36 columns representing different aspects of IT incidents.

These include:
- incident details such as `incident_state`, `priority`, and `category`
- time-related features such as `opened_at`, `resolved_at`, and `closed_at`
- assignment information such as `assignment_group` and `assigned_to`
- activity-related features such as `reassignment_count` and `reopen_count`

This shows that the dataset contains both categorical and numerical features, which will be useful for analysis and modeling.

## Data Type Analysis

In [7]:
incident_df.dtypes

number                     object
incident_state             object
active                       bool
reassignment_count          int64
reopen_count                int64
sys_mod_count               int64
made_sla                     bool
caller_id                  object
opened_by                  object
opened_at                  object
sys_created_by             object
sys_created_at             object
sys_updated_by             object
sys_updated_at             object
contact_type               object
location                   object
category                   object
subcategory                object
u_symptom                  object
cmdb_ci                    object
impact                     object
urgency                    object
priority                   object
assignment_group           object
assigned_to                object
knowledge                    bool
u_priority_confirmation      bool
notify                     object
problem_id                 object
rfc           

The data types of each column were examined to understand the structure of the dataset.

It was observed that some columns such as `reassignment_count`, `reopen_count`, and `sys_mod_count` are stored as numerical values, while columns like `active`, `made_sla`, and `knowledge` are stored as boolean values.

However, most of the columns are stored as object type. In particular, important time-related columns such as `opened_at`, `resolved_at`, and `closed_at` are stored as object instead of datetime format.

These columns need to be converted into datetime format to allow proper time-based analysis, such as calculating incident resolution time.

Additionally, the presence of many object-type columns indicates that the dataset contains a large number of categorical features, which may need to be encoded before applying machine learning models.

## Missing Value Analysis 

In [8]:
incident_df.isnull().sum().sort_values(ascending=False)

number                     0
incident_state             0
active                     0
reassignment_count         0
reopen_count               0
sys_mod_count              0
made_sla                   0
caller_id                  0
opened_by                  0
opened_at                  0
sys_created_by             0
sys_created_at             0
sys_updated_by             0
sys_updated_at             0
contact_type               0
location                   0
category                   0
subcategory                0
u_symptom                  0
cmdb_ci                    0
impact                     0
urgency                    0
priority                   0
assignment_group           0
assigned_to                0
knowledge                  0
u_priority_confirmation    0
notify                     0
problem_id                 0
rfc                        0
vendor                     0
caused_by                  0
closed_code                0
resolved_by                0
resolved_at   

In [9]:
(incident_df == "?").sum().sort_values(ascending=False)

caused_by                  141689
vendor                     141468
cmdb_ci                    141267
rfc                        140721
problem_id                 139417
sys_created_by              53076
sys_created_at              53076
u_symptom                   32964
assigned_to                 27496
assignment_group            14213
opened_by                    4835
resolved_at                  3141
closed_code                   714
resolved_by                   226
subcategory                   111
category                       78
location                       76
caller_id                      29
incident_state                  0
number                          0
reassignment_count              0
active                          0
contact_type                    0
sys_updated_by                  0
made_sla                        0
sys_mod_count                   0
reopen_count                    0
opened_at                       0
sys_updated_at                  0
u_priority_con

After checking for missing values using `isnull()`, it was found that there were no null values in the dataset. However, further analysis revealed that missing values were represented using the `"?"` symbol instead of actual null values.

A check was performed to count the number of `"?"` values in each column. The results showed that several columns contain a very high number of missing values.

Columns such as `caused_by`, `vendor`, `cmdb_ci`, `rfc`, and `problem_id` contain missing values in almost all rows (more than 99%).

Some columns such as `assigned_to`, `assignment_group`, and `u_symptom` contain a moderate number of missing values and may require further handling.

On the other hand, important columns such as `priority`, `impact`, `urgency`, and `incident_state` do not contain missing values and can be reliably used for analysis and modeling.

This step highlights the importance of checking for hidden missing values in real-world datasets.

## Understanding Incident Structure

In [10]:
incident_df["number"].value_counts().head(10)

number
INC0019396    58
INC0044260    56
INC0005927    46
INC0020718    45
INC0011206    44
INC0025734    43
INC0007349    43
INC0012815    40
INC0003419    38
INC0032450    38
Name: count, dtype: int64

In [11]:
incident_df["number"].nunique()

24918

To understand the structure of the dataset, the number of occurrences of each incident was checked.

It was found that the same incident number appears multiple times. For example, some incidents appear more than 50 times.

The dataset contains 24,918 unique incidents, while the total number of rows is much higher.

This shows that the dataset is not a simple one-row-per-incident dataset, but an event log, where each row represents an update or change in an incident.

This is important because it affects how the data will be prepared for modeling.

This means that a decision will be needed later on whether to use event-level data or convert it into one row per incident for modeling.

## Duplicate Row Check

In [12]:
incident_df.duplicated().sum()

np.int64(0)

The result showed that there are no exact duplicate rows in the dataset. This suggests that the repeated incident numbers are part of the event-log structure rather than accidental duplication.

## Initial Target Direction

Based on the dataset structure and the available time-related columns, the dataset is suitable for predicting incident resolution time.

Columns such as `opened_at`, `resolved_at`, and `closed_at` can be used to calculate incident duration.

This will be used later as the target variable for modeling in the data preparation stage.

## Resolution Time Availability

The `resolved_at` column contains some missing values.

This is important because if resolution time is used as the target variable, rows without a resolution timestamp cannot be used directly for target calculation.

These rows will need to be handled in the data preparation stage.